pallets/clusters/src/lib.rs

```diff
@@ -0,0 +1,479 @@
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn create_role(
+			origin: OriginFor<T>,
+			cluster_id: Hash128,
+			role_name: BoundedVec<u8, T::NameLimit>,
+			settings: RoleSetting,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!cluster_id.len().is_zero(), Error::<T>::InvalidInput);
+			ensure!(!role_name.len().is_zero(), Error::<T>::InvalidInput);
+			ensure!(!settings.name.len().is_zero(), Error::<T>::InvalidInput);
+			ensure!(!settings.data.len().is_zero(), Error::<T>::InvalidInput);
+
+			let role_hash = blake2_128(&[&cluster_id.as_slice(), role_name.as_slice()].concat());
+
+			// Check that the caller is the owner of the cluster
+			let cluster = <Clusters<T>>::get(&cluster_id).ok_or(Error::<T>::ClusterNotFound)?;
+			ensure!(who == cluster.owner, Error::<T>::NoPermission);
+
+			// At creation there are no Members and no Rules assigned to a Role
+			let new_role = Role { name: role_name.into_inner(), members: vec![], rules: None };
+
+			// Check that the role does not exists already in the cluster
+			if <Clusters<T>>::get(&cluster_id)
+				.iter()
+				.any(|cluster| cluster.roles.iter().any(|role| new_role.name.eq(&role.name)))
+			{
+				return Err(Error::<T>::RoleExists.into())
+			}
+
+			// write
+			<Clusters<T>>::mutate(&cluster_id, |cluster| {
+				let cluster = cluster.as_mut().unwrap();
+				cluster.roles.push(new_role);
+			});
+
+			<RoleToSettings<T>>::append(role_hash, settings);
+
+			Self::deposit_event(Event::RoleCreated { role_hash });
+			log::trace!("Role {:?} created and associated to cluster {:?}", role_hash, cluster_id);
+
+			Ok(())
+		}
+
+		/// Edit a **Role**.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1).ref_time())]
+		pub fn edit_role(
+			origin: OriginFor<T>,
+			role_name: BoundedVec<u8, T::NameLimit>,
+			cluster_id: Hash128,
+			new_members_list: BoundedVec<T::AccountId, T::MembersLimit>,
+			new_settings: RoleSetting,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!role_name.len().is_zero(), Error::<T>::InvalidInput);
+
+			if new_members_list.len().is_zero() &&
+				(new_settings.data.is_empty() || new_settings.name.is_empty())
+			{
+				return Err(Error::<T>::InvalidInput.into())
+			}
+
+			// Check that the caller is the owner of the cluster
+			let cluster = <Clusters<T>>::get(&cluster_id).ok_or(Error::<T>::SystematicFailure)?;
+			ensure!(who == cluster.owner, Error::<T>::NoPermission);
+
+			// Check that the role exists in the cluster and in storage
+			if !<Clusters<T>>::get(&cluster_id)
+				.iter()
+				.any(|cluster| cluster.roles.iter().any(|role| role_name.eq(&role.name)))
+			{
+				return Err(Error::<T>::RoleNotFound.into())
+			}
+
+			let role_hash = blake2_128(&[&cluster_id[..], &role_name.as_slice()].concat());
+			let role_name_vec = role_name.into_inner();
+
+			// write
+			<Clusters<T>>::mutate(&cluster_id, |cluster| {
+				let cluster = cluster.as_mut().unwrap();
```

Comment:
unwrap?

Reply:
At this point I am sure that this unwrap will not cause any problem. So it is safe to use it. We do this multiple times in all the other pallets too.